### ISSUE TITLE
Race condition with parallel testing creating rootdir 

### ISSUE THREAD

**Issue by kentfredric** [dagolden/Test-TempDir-Tiny#2]
This one is hard to reproduce, but I noticed this oddity in dzp MetaProvides::Package.

```
t/00-report-prereqs.t ......................................... ok
t/00-compile/lib_Dist_Zilla_Plugin_MetaProvides_Package_pm.t .. ok
File exists at /home/kent/perl5/perlbrew/perls/5.21.6/lib/site_perl/5.21.6/Test/TempDir/Tiny.pm line 78.
t/dz2-provides.t .............................................. 
Dubious, test returned 17 (wstat 4352, 0x1100)
No subtests run 
File exists at /home/kent/perl5/perlbrew/perls/5.21.6/lib/site_perl/5.21.6/Test/TempDir/Tiny.pm line 78.
t/finder-installmodules.t ..................................... 
Dubious, test returned 17 (wstat 4352, 0x1100)
No subtests run 
t/dz2-hidden-provides.t ....................................... ok
t/perl-5-14-package.t ......................................... ok
```

Code in question is 
```perl
       if ( -e $ROOT_DIR ) {
            croak("$ROOT_DIR is not a directory")
              unless -d $ROOT_DIR;
        }
        else {
            mkdir $ROOT_DIR or die $!;
        }
```

So it stands to reason, under parallel jobs both using Test::TempDir, one of those jobs could do
`-e $ROOT_DIR` , observe that the directory does not exist, but by the time it calls mkdir, some other thread has already created it.

```
1: -e $ROOTDIR => false
2: -e $ROOTDIR => false
1: mkdir $ROOTDIR => ok
2: mkdir $ROOTDIR => exists
```

**Comment by dagolden** [comment]
Great catch. The die needs to be conditional only if directory doesn't
exist.

**Comment by kentfredric** [comment]
I think there's scope for a potential similar bug to occur.

```
1. mkdir $ROOTDIR => ok
1. <test>
1. <tests finish>
2. assert $ROOTDIR exists => ok
1. Test OK, cleanup $ROOTDIR
2. add file to $ROOTDIR => BANG, no $ROOTDIR.
```
Not sure how to fix that. Every time I tried to codify a fix for the initial problem I kept feeling like I was doing something wrong.

**Comment by dagolden** [comment]
Yeah. It probably has to detect parallel operation and not clean up.  Or
everything has to come up to the distribution root directory.

Or some sort of more complicated semaphore system and retry on directory
creation failure.

Distributed systems are hard. :-/

**Comment by kentfredric** [comment]
And now we know why Test::Tempdir had that horrible NFSLock stuff

```
lock files are used on t/tmp, to prevent race conditions when running under a parallel test harness.
```

**Comment by kentfredric** [comment]
There's a plausible way it could side step this problem: 

Eliminate the 2-layer structure and simply use a prefix notation instead.

That is, instead of:

 ```
t/tmp/foo
t/tmp/bar
```
Simply have
```
t/tmp-foo
t/tmp-bar
```

This solves the problem by eliminating a common path and each can continue to clean up after itself.

And using a prefix means it should still be straight forward to identify such folders and add ignore rules.

That would incidentally make the `_init` code cleaner, because you could eliminate the second layer of indirection that you inherently have when `$ROOT` is a /tmp/<random> dir, so the layers would be

```
/tmp/RANDOM
/tmp/RANDOM
```
instead of
```
/tmp/RANDOM/foo
/tmp/RANDOM/bar
```

**Comment by dagolden** [comment]
I'd prefer to keep the nested directory structure for cleanliness.

The most common race -- having ./tmp already existing, is easy to fix.
The more subtle race of having it delete out from underneath requires
some retries, which probabilistically should eventually succeed at
the levels of parallelism we usually see for testing.

Please check out b435a37e5aafcb8769d1a3a3bfa81ffa6f0cc8de which implements
this as well as a fall back to regular tempfiles and tell me if you think
it seems sane.

**Comment by kentfredric** [comment]
The approach and code look reasonable. Obviously its going to be hard to know if it solves the problem or not, given how difficult it its to create it in the first place :) 

I was fortunately the first one to stumble into it of [all the places I use it so far!](https://metacpan.org/requires/distribution/Dist-Zilla-Util-Test-KENTNL?sort=[[2,1]]&size=500)

(Granted the code in question only lazy-uses it if its installed for convenience, narrowing the numbers of places it could possibly fail substantially )

**Comment by dagolden** [comment]
Shipped 0.003 to CPAN.

I also did a separate test with 1000 .t files running in parallel.  I
was able to provoke the error on the first run with 0.002 but was unable
to do so after several runs with 0.003.  This doesn't prove it can't
happen, but at least suggests it's far less likely.
